Adding JsonView decorator

snirs90 opened the pull request:
Adding @JsonView decorator so now it'll be possible to annotate a method in a controller class to expose the object properties.

Similar to JMS Serializer in Symfony (PHP) and in Java Spring.

Very useful with a library such - "ts-express-decorators" (https://github.com/Romakita/ts-express-decorators)

So it is more simpler to return a serialized object.

An example:

```
@Exclude()
class UserModel {
    @Expose()
    id: number;

    @Expose({name: 'first_name', groups: ['user.base'] })
    firstName: string;

    @Expose({ name: 'last_name', groups: ['user.base'] })
    lastName: string;

    @Expose({ groups: ['user.details'] })
    email: string;

    phone: string;
}

@Controller("/users")
export class UserController {

    constructor() {}

    @Get("/:id")
    @JsonView({groups: ['user.base']})
    public getDetails(
        @Request() request: Express.Request, 
        @Response() response: Express.Response,
        @Next() next: Express.NextFunction
    ): any {

        let user = new UserTest();
        user.id = 1;
        user.firstName = 'Snir';
        user.lastName = 'Segal';
        user.email = 'snirs@codeoasis.com';
        user.phone = '0523262583';

        return user;
    }
}
```

pleerock commented on test/functional/basic-functionality.spec.ts:
```diff
@@ -1680,5 +1680,47 @@ describe("basic functionality", () => {
+            firstName: "Umed",
+            lastName: "Khudoiberdiev"
+        };
+
```
better to move `user` and `plainUser` block into controller action code to avoid extra thinking about their scope and why they are defined here.

pleerock commented on src/decorators.ts:
```diff
@@ -48,4 +49,24 @@ export function Exclude(options?: ExcludeOptions) {
+/**
+ * Return the object with the exposed properties only.
+ */
+export function JsonView(params?: {}, method?: string): Function {
```
`params?: {}` should be strict typed. Added properties and their types.

pleerock replied:
`method?: string` should be a string literal type, e.g. `"classToPlain"|"plainToClass"|...`. This will make it type-safe, predictable and prevent you doing `const classTransformer: any` below.

pleerock replied:
maybe its good idea to move `method` inside params.

pleerock commented on src/decorators.ts:
```diff
@@ -48,4 +49,24 @@ export function Exclude(options?: ExcludeOptions) {
+
+        descriptor.value = function(...args: any[]) {
+            let transformer: Function = typeof method === "string" && method ? classTransformer[method] : classTransformer.classToPlain;
+            let result: any = originalMethod.apply(this, args);
```
once method will be a string literal type you can use simple switch-cases to make this code both more readable and understandable + make it more typesafe.

pleerock commented on test/functional/basic-functionality.spec.ts:
```diff
@@ -1680,5 +1680,47 @@ describe("basic functionality", () => {
+        expect(result.password).to.be.undefined;
+        expect(result).to.be.eql(plainUser);
+    });
 
```
please add more complete test coverage for using `@JsonView` with all params it have.

pleerock commented on src/decorators.ts:
```diff
@@ -48,4 +49,24 @@ export function Exclude(options?: ExcludeOptions) {
+            let result: any = originalMethod.apply(this, args);
+            
+            let isPromise = !!result && (typeof result === "object" || typeof result === "function") && typeof result.then === "function";
+
```
use `const` everywhere possible.

pleerock commented on src/decorators.ts:
```diff
@@ -48,4 +49,24 @@ export function Exclude(options?: ExcludeOptions) {
+            let transformer: Function = typeof method === "string" && method ? classTransformer[method] : classTransformer.classToPlain;
+            let result: any = originalMethod.apply(this, args);
+            
+            let isPromise = !!result && (typeof result === "object" || typeof result === "function") && typeof result.then === "function";
```
there is a `PromiseUtils.isPromise()` method, you can use that instead of this code.

snirs90 replied:
I'm not aware of this method in NodeJS nor in JavaScript

pleerock commented:
Thank you for this PR. Love your idea. I'm using a bit alternative approach in my [routing-controllers](https://github.com/pleerock/routing-controllers) library. What you did will perfectly fit in this library too. Please fix all issues I have mentioned before it can be merged. Also please add docs about this feature.

pleerock commented:
Also TBH I don't like `@JsonView` name. Do you have a better naming suggestions?

snirs90 commented:
Hi,
Thanks for reviewing, I'll work on the fixes.

About the annotation naming -
Maybe `@ResponseJson` or `@ResponseTransform` ?

pleerock commented:
The process itself is not related to request/response anyhow, but can be used in all use cases including controllers with response. Maybe something like `@TransformMethod` ? Because it transforms method result?

snirs90 commented:
Hi,
I've fixed all the issues you mentioned, added more tests and documentation.

Changed the decorator name to `@TransformMethod`, although I'm still trying to think for a better name.
If you find any other issues please tell me about it.

The decorator supports only `classToPlain` and `classToClass`, I don't see any reason to use plainToClass/serialize/deserialize for this decorator at the moment.

If you think there is important usage for those I will add support for them also.

snirs90 commented:
@pleerock any updates?

Thanks,

pleerock commented on test/functional/transformer-method.spec.ts:
```diff
@@ -0,0 +1,223 @@
+
+        class UserController {
+            
+            @TransformMethod({}, "classToClass")
```
this does not look good

pleerock commented:
I was thinking about naming of this decorator... Maybe we should remove "method" param and create two decorators instead?  

So instead of:

`TransformMethod(params?: ClassTransformOptions, method?: "classToPlain"|"classToClass")`

do:

`ClassToPlainMethodResult(params?: ClassTransformOptions)`
`ClassToClassMethodResult(params?: ClassTransformOptions)`

or maybe?: 

`ClassToPlainMethod(params?: ClassTransformOptions)`
`ClassToClassMethod(params?: ClassTransformOptions)`


which is less correct but more pretty, not sure which one is better

snirs90 commented:
@pleerock I've changed it to two decorators:
`@TransformClassToPlain` and `@TransformClassToClass`
I think now the names are good and clear enough.

Let me know if you think of another things that we should add.

pleerock commented on README.md:
````diff
@@ -590,6 +590,55 @@ Now when you call `plainToClass` and send a plain representation of the Photo ob
+
+An example:
+
+```
````
should be ```typescript

pleerock commented on src/decorators.ts:
```diff
@@ -48,4 +50,42 @@ export function Exclude(options?: ExcludeOptions) {
+        
+        descriptor.value = function(...args: any[]) {
+            const result: any = originalMethod.apply(this, args);
+            const isPromise = !!result && (typeof result === "object" || typeof result === "function") && typeof result.then === "function";          
```
there is something like `PromiseUtils.isPromise` in the lib that you can use

snirs90 replied:
In which lib?

pleerock commented on src/decorators.ts:
```diff
@@ -48,4 +50,42 @@ export function Exclude(options?: ExcludeOptions) {
+
+    return function (target: Function, propertyKey: string, descriptor: PropertyDescriptor) {
+        const classTransformer: ClassTransformer = new ClassTransformer();
+        const MethodTransformer: Function = classTransformer.classToPlain;
```
`MethodTransformer` is redundant here. Just call `classTransformer.classToPlain` below.

pleerock commented:
I think names are good. Please update the docs to reflect latest changes + small comments I have left that applies to both decorators. Also add exports to `index.ts`.

pleerock commented:
btw, thank you for this feature. Idea of how implement it is really cool and I like it so much

snirs90 commented:
@pleerock I've made the changes.
About `PromiseUtils.isPromise` maybe you got confused with another project?
I'm not aware of any JavaScript method which checks if a variable is a promise.

pleerock commented:
> About PromiseUtils.isPromise

ah yeah, sorry about that, I thought I have this function in this lib too

pleerock commented:
thank you! released in `0.1.4`
